### PR TITLE
Add OpenSearch support to Adstash using an extra flag

### PR DESCRIPTION
**Pull request by jamesacris**
At STFC (Science & Technology Facilities Council) we are migrating from Elasticsearch to OpenSearch for our search clusters. This PR adds OpenSearch support to Adstash so we can use it to send job logs from our HTCondor system to our search cluster.

### CODE REVIEW

**Comment by jamesacris** [comment]
Just a note to say this is still a WIP. Help is appreciated as I'm not sure how to deal with conditionally importing the right elasticsearch/opensearch module, and propogate the argument properly through the codebase.

**Comment by jamesacris** [comment]
This is now working but needs refactoring. I based it on the codebase from a few commits ago https://github.com/htcondor/htcondor/tree/0e2a42f2201a1ce7b50069d963f77713d2bd0319/src/condor_scripts/adstash as the next commit https://github.com/htcondor/htcondor/tree/b38ff8838cd31243b2ea0bb2ac1f6f0e027c13d7/src/condor_scripts/adstash seemed to break things for me.

File: src/condor_scripts/adstash/elastic.py [hunk 1]
```diff
@@ -183,10 +159,16 @@ def __init__(
...
         use_https=False,
+        use_opensearch=False,
         ca_certs=None,
         logdir=Path.cwd(),
     ):
 
+        if use_opensearch:
+            import opensearchpy as elasticsearch
+        else:
+            import elasticsearch
```
Comment (jamesacris) [inline src/condor_scripts/adstash/elastic.py:170]: Imports in the middle of a file are definitely not ideal

File: src/condor_scripts/adstash/elastic.py [hunk 2]
```diff
@@ -212,12 +194,23 @@ def __init__(
...
 
-        self.handle = elasticsearch.Elasticsearch([es_client])
-        super().__init__(self, logdir)
+        if use_opensearch:
+            self.use_opensearch = True
+            self.handle = elasticsearch.OpenSearch([es_client])
+        else:
+            self.use_opensearch = False
+            self.handle = elasticsearch.Elasticsearch([es_client])
+        self.logdir = Path(logdir)
```
Comment (jamesacris) [inline src/condor_scripts/adstash/elastic.py:205]: Imports in the middle of a file are definitely not ideal (2 of 2)

File: src/condor_scripts/adstash/elastic.py [hunk 3]
```diff
@@ -212,12 +194,23 @@ def __init__(
                 logging.error("Using HTTPS with Elasticsearch requires that either ca_certs be provided or certifi library be installed")
                 sys.exit(1)
             es_client["use_ssl"] = True
-            es_client["verify_certs"] = True
+            es_client["verify_certs"] = False
 
-        self.handle = elasticsearch.Elasticsearch([es_client])
-        super().__init__(self, logdir)
+        if use_opensearch:
+            self.use_opensearch = True
```
Comment (jamesacris) [inline src/condor_scripts/adstash/elastic.py:200]: I had to give the ElasticInterface class a new attribute in order to use the use_opensearch argument in make_mapping() on line 207

File: src/condor_scripts/adstash/elastic.py [hunk 4]
```diff
@@ -212,12 +194,23 @@ def __init__(
...
+            self.use_opensearch = False
+            self.handle = elasticsearch.Elasticsearch([es_client])
+        self.logdir = Path(logdir)
 
     def make_mapping(self, idx):
+
+        if self.use_opensearch:
+            import opensearchpy as elasticsearch
+        else:
+            import elasticsearch
```
Comment (jamesacris) [inline src/condor_scripts/adstash/elastic.py:212]: Imports in the middle of a file are definitely not ideal (2 of 2)

File: src/condor_scripts/adstash/history.py [hunk 5]
```diff
@@ -385,6 +300,10 @@ def process_histories(
     Process history files for each schedd listed in a given
     multiprocessing pool
     """
+    if args.es_use_opensearch:
+        import opensearchpy as elasticsearch
+    else:
+        import elasticsearch
```
Comment (jamesacris) [inline src/condor_scripts/adstash/history.py:306]: Imports in the middle of a file are definitely not ideal (3 of 3)

**Comment by jasoncpatton** [comment]
@jamesacris I'm in the middle of refactoring adstash in prep for HTCondor 10.0 to make adding new interfaces more straightforward. If I'm able to get to adding the OpenSearch interface, I'll try to pull some of these commits so you get credited with the contribution.

**Comment by jamesacris** [comment]
@jasoncpatton That's good to hear - I will leave it in your hands! Thanks very much.

**Pull request closed without merging**